File: Squalr.Engine.Scanning/Snapshots/SnapshotElementRange.cs
namespace Squalr.Engine.Scanning.Snapshots
{
    using Squalr.Engine.Common;
    using System;
    using System.Collections.Generic;

    /// <summary>
    /// Defines a range of snapshot elements in an external process.
    /// </summary>
    public class SnapshotElementRange
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="SnapshotElementRange" /> class.
        /// </summary>
        /// <param name="parentRegion">The parent region of this snapshot element range.</param>
        public SnapshotElementRange(SnapshotRegion parentRegion) : this(parentRegion, 0, parentRegion?.RegionSize ?? 0)
        {
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="SnapshotElementRange" /> class.
        /// </summary>
        /// <param name="parentRegion">The parent region of this snapshot element range.</param>
        /// <param name="regionOffset">The base address of this snapshot region.</param>
        /// <param name="range">The size of this snapshot region.</param>
        public SnapshotElementRange(SnapshotRegion parentRegion, Int32 regionOffset, Int32 range)
        {
            this.ParentRegion = parentRegion;
            this.RegionOffset = regionOffset;
            this.Range = range;
        }

        /// <summary>
        /// Gets the snapshot region from which this element range reads its values.
        /// </summary>
        public SnapshotRegion ParentRegion { get; private set; }

        /// <summary>
        /// Gets the offset from the base of the snapshot region that contains this element range.
        /// </summary>
        public Int32 RegionOffset { get; private set; }

        /// <summary>
        /// Gets the size of this element range in bytes. This is the number of bytes directly contained, but more bytes may be used if tracking data types larger than 1-byte.
        /// </summary>
        public Int32 Range { get; private set; }

        /// <summary>
        /// Gets the address of the first element contained in this snapshot region.
        /// </summary>
        public UInt64 BaseElementAddress
        {
            get
            {
                return unchecked(this.ParentRegion.BaseAddress + (UInt64)this.RegionOffset);
            }
        }

        /// <summary>
        /// Gets the address of the last element contained in this snapshot region (assuming 1-byte alignment).
        /// </summary>
        public UInt64 EndElementAddress
        {
            get
            {
                return unchecked(this.ParentRegion.BaseAddress + (UInt64)(this.RegionOffset + this.Range));
            }
        }

        /// <summary>
        /// Gets or sets the base index of this snapshot element, relative to the parent snapshot region base index. Used for indexing scan results.
        /// </summary>
        public Int32 SnapshotRegionRelativeIndex { get; set; }

        /// <summary>
        /// Gets the size of this range in bytes. This requires knowing what data type is being tracked, since data types larger than 1 byte will overflow out of this region.
        /// Also, this takes into account how much space is available for reading from the snapshot region containing this element range.
        /// </summary>
        /// <param name="dataTypeSize">The data type size of the elements contained.</param>
        /// <returns>The true byte count of this range given a data type.</returns>
        public Int32 GetByteCount(Int32 dataTypeSize)
        {
            Int32 desiredSpillOverBytes = Math.Max(dataTypeSize - 1, 0);
            Int32 availableSpillOverBytes = unchecked((Int32)(this.ParentRegion.EndAddress - this.EndElementAddress));
            Int32 usedSpillOverBytes = Math.Min(desiredSpillOverBytes, availableSpillOverBytes);

            return this.Range + usedSpillOverBytes;
        }

        /// <summary>
        /// Gets the number of elements contained in this snapshot.
        /// <param name="alignment">The memory address alignment of each element.</param>
        /// </summary>
        public Int32 GetAlignedElementCount(MemoryAlignment alignment)
        {
            Int32 alignmentValue = unchecked((Int32)alignment);
            Int32 elementCount = this.Range / (alignmentValue <= 0 ? 1 : alignmentValue);

            return elementCount;
        }

        /// <summary>
        /// Resize the snapshot region for safe reading given an allowed data type size.
        /// </summary>
        /// <param name="dataTypeSize"></param>
        public void ResizeForSafeReading(Int32 dataTypeSize)
        {
            Int32 parentRegionSize = this.ParentRegion?.RegionSize ?? 0;

            this.Range = Math.Clamp(this.Range, 0, parentRegionSize - this.RegionOffset - dataTypeSize);
        }

        /// <summary>
        /// Indexer to allow the retrieval of the element at the specified index.
        /// </summary>
        /// <param name="index">The index of the snapshot element.</param>
        /// <returns>Returns the snapshot element at the specified index.</returns>
        public SnapshotElementIndexer this[Int32 index, MemoryAlignment alignment]
        {
            get
            {
                return new SnapshotElementIndexer(elementRange: this, elementIndex: index, alignment: alignment);
            }
        }

        /// <summary>
        /// Gets the enumerator for an element reference within this snapshot region.
        /// </summary>
        /// <returns>The enumerator for an element reference within this snapshot region.</returns>
        public IEnumerator<SnapshotElementIndexer> IterateElements(MemoryAlignment alignment)
        {
            Int32 elementCount = this.GetAlignedElementCount(alignment);
            SnapshotElementIndexer snapshotElement = new SnapshotElementIndexer(elementRange: this, alignment: alignment);

            for (snapshotElement.ElementIndex = 0; snapshotElement.ElementIndex < elementCount; snapshotElement.ElementIndex++)
            {
                yield return snapshotElement;
            }
        }
    }
    //// End class
}
//// End namespace

File: squalr-engine-scanning/src/snapshots/snapshot_element_range.rs
use crate::snapshots::snapshot_element_indexer::SnapshotElementIndexer;
use crate::snapshots::snapshot_region::SnapshotRegion;
use squalr_engine_memory::memory_alignment::MemoryAlignment;
use std::iter::Iterator;
use std::sync::{Arc, RwLock};

#[derive(Debug, Clone)]
pub struct SnapshotElementRange {
    pub parent_region: Arc<RwLock<SnapshotRegion>>,
    region_offset: usize,
    range: usize,
}

impl SnapshotElementRange {
    pub fn new(parent_region: Arc<RwLock<SnapshotRegion>>) -> Self {
        let parent_region_size = parent_region.read().unwrap().get_region_size() as usize;
        Self::with_offset_and_range(parent_region, 0, parent_region_size)
    }

    pub fn with_offset_and_range(parent_region: Arc<RwLock<SnapshotRegion>>, region_offset: usize, range: usize) -> Self {
        Self {
            parent_region,
            region_offset,
            range,
        }
    }

    pub fn get_base_element_address(&self) -> u64 {
        return self.parent_region.read().unwrap().get_base_address() + self.region_offset as u64;
    }

    pub fn get_end_element_address(&self) -> u64 {
        return self.get_base_element_address() + self.range as u64;
    }

    pub fn get_region_offset(&self) -> usize {
        return self.region_offset;
    }
    
    pub fn get_range(&self) -> usize {
        return self.range;
    }

    pub fn get_byte_count(&self, data_type_size: usize) -> usize {
        let desired_spill_over_bytes = if data_type_size > 1 { data_type_size - 1 } else { 0 };
        let available_spill_over_bytes = (self.parent_region.read().unwrap().get_base_address() + self.range as u64) - self.get_end_element_address();
        let used_spill_over_bytes = std::cmp::min(desired_spill_over_bytes, available_spill_over_bytes as usize);

        return self.range + used_spill_over_bytes;
    }

    pub fn get_aligned_element_count(&self, alignment: MemoryAlignment) -> usize {
        let alignment_value = alignment as usize;

        return self.range / if alignment_value == 0 { 1 } else { alignment_value };
    }

    pub fn resize_for_safe_reading(&mut self, data_type_size: usize) {
        let parent_region_size = self.parent_region.read().unwrap().get_region_size() as usize;
        self.range = std::cmp::min(self.range, parent_region_size - self.region_offset - data_type_size);
    }

    pub fn get_element_indexer(&self, index: usize, alignment: MemoryAlignment) -> SnapshotElementIndexer {
        return SnapshotElementIndexer::new(self.clone().into(), alignment, index);
    }

    pub fn iterate_elements(&self, alignment: MemoryAlignment) -> SnapshotElementIterator {
        return SnapshotElementIterator::new(self.clone().into(), alignment);
    }
}

pub struct SnapshotElementIterator {
    element_range: Arc<SnapshotElementRange>,
    alignment: MemoryAlignment,
    current_index: usize,
    total_elements: usize,
}

impl SnapshotElementIterator {
    pub fn new(element_range: Arc<SnapshotElementRange>, alignment: MemoryAlignment) -> Self {
        let total_elements = element_range.get_aligned_element_count(alignment);
        Self {
            element_range,
            alignment,
            current_index: 0,
            total_elements,
        }
    }
}

impl Iterator for SnapshotElementIterator {
    type Item = SnapshotElementIndexer;

    fn next(&mut self) -> Option<Self::Item> {
        if self.current_index < self.total_elements {
            let indexer = self.element_range.get_element_indexer(self.current_index, self.alignment);
            self.current_index += 1;

            return Some(indexer);
        } else {
            return None;
        }
    }
}
